libraries/lib-graphics/graphics/Rect.h

```diff
@@ -0,0 +1,267 @@
+
+   return { p31, SizeType<DataType> { p32.x - p31.x, p32.y - p32.y - p31.y } };
+}
+
+template <typename DataType>
+bool operator==(
+   const RectType<DataType> lhs, const RectType<DataType> rhs) noexcept
+{
+   return lhs.origin == rhs.origin && lhs.size == rhs.size;
+}
+
+template <typename DataType>
+bool operator!=(
+   const RectType<DataType> lhs, const RectType<DataType> rhs) noexcept
+{
+   return !(lhs == rhs);
+}
+
+//! Cast the rectangle to another data type.
+template <typename To, typename From>
+RectType<To> rect_cast(RectType<From> rect)
+{
+   return { point_cast<To>(rect.origin), size_cast<To>(rect.size) };
+}
+
+//! Alias for rectangle with float data type.
+using Rect = RectType<float>;
+
+//! Axis aligned bounding box.
+template<typename DataType>
+struct AxisAlignedBoundingBoxType
+{
+   //! Top left corner of the box
+   PointType<DataType> topLeft { std::numeric_limits<DataType>::max(),
+                                 std::numeric_limits<DataType>::max() };
+
+   //! Bottom right corner of the box
+   PointType<DataType> bottomRight { std::numeric_limits<DataType>::min(),
+                                     std::numeric_limits<DataType>::min() };
+
+   //! Return true if the box is not empty.
+   bool IsValid() const noexcept
+   {
+      return topLeft.x <= bottomRight.x && topLeft.y <= bottomRight.y;
+   }
+
+   AxisAlignedBoundingBoxType() = default;
+
+   //! Constructs the box from the specified point.
+   explicit AxisAlignedBoundingBoxType(PointType<DataType> point)
+       : topLeft(point)
+       , bottomRight(point)
+   {
+   }
+
+   //! Constructs the box from the specified rectangle.
+   explicit AxisAlignedBoundingBoxType(RectType<DataType> rect)
+       : topLeft(rect.origin)
+       , bottomRight(rect.origin + PointType<DataType>{ rect.size.width, rect.size.height })
+   {
+   }
+
+   //! Constructs the box from the specified points.
+   AxisAlignedBoundingBoxType(PointType<DataType> p1, PointType<DataType> p2)
+       : topLeft { std::min(p1.x, p2.x), std::min(p1.y, p2.y) }
+       , bottomRight { std::max(p1.x, p2.x), std::max(p1.y, p2.y) }
+   {
+   }
+
+   AxisAlignedBoundingBoxType(const AxisAlignedBoundingBoxType &) = default;
+   AxisAlignedBoundingBoxType(AxisAlignedBoundingBoxType &&) = default;
+   AxisAlignedBoundingBoxType &operator=(const AxisAlignedBoundingBoxType &) = default;
+   AxisAlignedBoundingBoxType &operator=(AxisAlignedBoundingBoxType &&) = default;
+
+   //! Expands the box to include the specified point.
+   AxisAlignedBoundingBoxType& Expand(PointType<DataType> point) noexcept
+   {
+      topLeft.x = std::min(topLeft.x, point.x);
+      topLeft.y = std::min(topLeft.y, point.y);
+      bottomRight.x = std::max(bottomRight.x, point.x);
+      bottomRight.y = std::max(bottomRight.y, point.y);
+
+      return *this;
+   }
+
+   //! Expands the copy of the box to include the specified rectangle.
+   AxisAlignedBoundingBoxType Expanded(PointType<DataType> point) const noexcept
```

Comment:
I might name it with a noun, `Expansion`

Reply:
I have a strong opinion that function and methods must be verbs. I will push hard towards it.